crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -0,0 +1,231 @@
+use self::widget::Widget;
+use super::{
+    handler::{self, Capabilities, OpenIDState},
+    messages::{
+        capabilities::{Filter, Options, EventFilter},
+        from_widget::{ReadEventRequest, ReadEventResponse, SendEventRequest, SendEventResponse},
+        {openid, MatrixEvent},
+    },
+    {Error::WidgetError, Result},
+};
+use crate::room::MessagesOptions;
+use crate::{event_handler::EventHandlerHandle, room::Joined};
+use async_trait::async_trait;
+use ruma::{api::client::filter::RoomEventFilter, events::AnySyncTimelineEvent, serde::Raw};
+use tokio::sync::mpsc;
+
+pub mod widget;
+
+#[derive(Debug)]
+pub struct Driver<W: Widget> {
+    pub matrix_room: Joined,
+    pub widget: W,
+    add_event_handler_handle: Option<EventHandlerHandle>,
+}
+
+#[async_trait(?Send)]
+impl<W: Widget> handler::Driver for Driver<W> {
+    async fn send(&self, _message: handler::Outgoing) -> Result<()> {
+        // let message_str = serde_json::to_string(&message)?;
+        let _ = self.widget.send_widget_message("TODO get message string from outgoing");
+        Result::Ok(())
+    }
+    async fn initialise(&mut self, options: Options) -> Result<Capabilities> {
+        let options = self.widget.capability_permissions(options).await?;
+
+        let mut capabilities = Capabilities::default();
+
+        capabilities.event_listener = self.build_event_listener(&options.read_filter);
+        capabilities.event_sender = self.build_event_sender(&options.send_filter);
+        capabilities.event_reader = self.build_event_reader(&options.read_filter);
+        Result::Ok(capabilities)
+    }
+    async fn get_openid(&self, req: openid::Request) -> OpenIDState {
+        // TODO: make the client ask the user first.
+        // We wont care about this for Element call -> Element X
+        // if !self.has_open_id_user_permission() {
+        //     let (rx,tx) = tokio::oneshot::channel();
+        //     return OpenIDState::Pending(tx);
+        //     let permissions = widget.openid_permissions().await?;
+        //     if (permissions.should_cache){
+        //        self.auto_allow_openid_token(permission.is_allowed);
+        //     }
+        //     self.get_openid(req, Some(tx)); // get open id can be called with or without tx and either reutrns as return or sends return val over tx
+        // }
+
+        let user_id = self.matrix_room.client.user_id();
+        if user_id == None {
+            return OpenIDState::Resolved(Err(WidgetError(
+                "Failed to get an open id token from the homeserver. Because the userId is not available".to_owned()
+            )));
+        }
+        let user_id = user_id.unwrap();
```

Comment:
A more elegant way of writing it is:

```suggestion
let Some(user_id) = self.matrix_room.client.user_id() else {
    return OpenIDState::Resolved(Err(WidgetError("No user ID available".to_owned())));
};
```

Note, however, that it looks like it's too late for checking for a user ID. It feels like it must be set prior to any other Widget API interactions unless we allow non-authenticated users to interface with the Widget API (which I think is not the case?).

Reply:
so there is
```
if let EnumCase(val) {
  val.can_only_be_accessed_in_this_scope();
}
```
```
let EnumCase(val) else {
  cannot_use_val_here();
}
val.can_be_accessed_in_this_scope();
```

Reply:
that is new. but very nice!